[Mind-7] 전문가 등록 기능 구현

TaskerJang opened the pull request:
# **Pull Request**

## 💡 PR 요약
전문가 등록 기능을 구현하고, `UserServiceTest`를 리팩토링하였습니다.

## 🔍 주요 변경사항
### **1. 전문가 등록 기능 추가**
- `Expert` 엔티티 및 관련 DTO (`ExpertCreateRequest`, `ExpertResponse`) 추가
- `ExpertController`, `ExpertService`, `ExpertRepository` 구현
- `ExpertServiceTest` 단위 테스트 작성

### **2. `UserServiceTest` 리팩토링**
- `UserServiceTest` 위치 이동 (`src/test/java/com/mindgoal/backend/user/service/`)
- 테스트 코드 가독성 개선 및 코드 리팩토링

## 🔗 연관된 이슈
- close #이슈번호

## 📸 스크린샷 (선택)
(필요하면 스크린샷 첨부)

## ✅ 체크리스트
- [x] 테스트 코드를 작성하였나요?
- [x] 관련 문서를 업데이트하였나요?
- [ ] Breaking Change가 있나요?
- [x] 코드 포맷팅과 린트 검사를 완료하였나요?

## 🙏 리뷰어 참고사항
- `ExpertService` 로직이 적절한지 확인 부탁드립니다.
- `UserServiceTest` 리팩토링 방향이 적절한지 검토 부탁드립니다.

## 📋 추가 컨텍스트 (선택)
- 전문가 등록 기능에 대한 추가적인 개선 사항이 있다면 의견 부탁드립니다.

Commit:
[refactor]: UserServiceTest 리팩토링

- UserServiceTest 위치 이동
- 테스트 코드 개선 및 가독성 향상

Commit:
[feat]: 전문가 등록 기능 구현

- Expert 엔티티 수정
- ExpertController, Service, Repository 구현
- Expert DTO 추가 (ExpertCreateRequest, ExpertResponse)
- Expert 관련 단위 테스트 추가 (ExpertServiceTest)

hwan2-99 commented on src/main/java/com/mindgoal/domain/expert/controller/ExpertController.java:
```diff
@@ -1 +1,28 @@
+        ExpertResponse response = expertService.register(request);
+        return ResponseEntity
+                .status(HttpStatus.CREATED)
+                .body(BaseResponse.success(response, "전문가 등록 성공"));
```
전문가 등록성공이라는 값에 대해서 따로 enum 클래스를 통해 관리해서 넣어주는게 더 좋을것같습니다!!

TaskerJang replied:
네, 좋은 의견 감사합니다! BaseResponseStatus에 관련 상태를 추가하여 반영했습니다:
- EXPERT_REGISTER_SUCCESS(true, 201, "전문가 등록이 성공했습니다") 추가
- 성공/에러 상태 코드와 메시지를 일관성있게 관리할 수 있게 되었습니다.

hwan2-99 commented on src/main/java/com/mindgoal/domain/expert/service/ExpertService.java:
```diff
@@ -1 +1,52 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional(readOnly = true)
```
transactional 레벨을 선언하신 이유가 궁금합니다! 물론 이렇게 하면 readOnly가 필요한 메서드들에는 어노테이션을 안써도되는 장점이있지만, 아래 register 클래스와 충돌할수도 있을것같아서요!!

TaskerJang replied:
지적해주신 대로 클래스 레벨의 @Transactional(readOnly = true)가 메서드 레벨의 @Transactional과 혼란을 줄 수 있겠네요.
해당 어노테이션을 제거하고 메서드 단위로 필요한 곳에만 적용하도록 하겠습니다.

hwan2-99 commented on src/main/java/com/mindgoal/domain/expert/service/ExpertService.java:
```diff
@@ -1 +1,52 @@
+
+            // 이미 전문가로 등록된 사용자인지 확인
+            if (expertRepository.existsByUserId(currentUser.getId())) {
+                throw new IllegalArgumentException("이미 등록된 전문가입니다.");
```
추후에 각도메인별로 예외처리 커스텀 클래스를 사용해서 명확한 예외를하는 방법을 회의하면 좋을것같습니다!

TaskerJang replied:
넵 알겠습니다!

hwan2-99 commented on src/main/java/com/mindgoal/domain/expert/service/ExpertService.java:
```diff
@@ -1 +1,52 @@
+                    .instagramUrl(request.getInstagramUrl())
+                    .isActive(true)
+                    .build();
+
```
해당 Expert를 dto를 통해 매핑해주는 부분은 private 메서드로 아래에 분리해서 호출하는 방식으로 하면 가독성이 더 높아질 것 같아요

TaskerJang replied:
네, Expert 엔티티 생성 로직을 createExpertEntity라는 private 메서드로 분리했습니다.

hwan2-99 reviewed (commented):
고생하셨습니다!

Commit:
[refactor]: PR 리뷰 반영

- BaseResponseStatus에 Expert 관련 상태 추가
- Expert 엔티티 생성 로직 분리
- ExpertService에 CustomException 적용

hwan2-99 reviewed (approved):
리뷰반영 좋습니다!! 고생하셨습니다.

justin8491 reviewed (approved):
전에 구현해본 방식과는 달라 학습하며 보고있습니다 객체 생성을 빌더로 작업한 이유가 있는지 궁금합니다 ㅎㅎ
수고하셨습니다.
